src/Features/CSharp/Portable/Testing/CSharpTestMethodFinder.cs

```diff
@@ -0,0 +1,51 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Features.Testing;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.CSharp.Testing;
+
+[ExportLanguageService(typeof(ITestMethodFinder), LanguageNames.CSharp), Shared]
+[method: ImportingConstructor]
+[method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+internal class CSharpTestMethodFinder([ImportMany] IEnumerable<ITestFrameworkMetadata> testFrameworks) : AbstractTestMethodFinder<MethodDeclarationSyntax>(testFrameworks)
+{
+    protected override async Task<bool> IsTestMethodAsync(Document document, MethodDeclarationSyntax method, CancellationToken cancellationToken)
+    {
+        var attributes = method.AttributeLists.SelectMany(a => a.Attributes);
+        foreach (var attribute in attributes)
+        {
+            var isTestAttribute = await IsTestAttributeAsync(document, attribute, cancellationToken).ConfigureAwait(false);
+            if (isTestAttribute)
+            {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    private async Task<bool> IsTestAttributeAsync(Document document, AttributeSyntax attribute, CancellationToken cancellationToken)
+    {
+        var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
This does do a semantic operation to get the code lens - however it seemed reasonable to me for a couple reasons
1.  We shouldn't be doing anything the majority of the time.  This only kicks in if there is an attribute on a method.
2.  We have to do semantic checks anyway when running the tests.

Happy to hear other thoughts though

Reply:
you're calling this for each attribute.  can you instead lift this out and pass it along.  in the current form, it would create/throwaway the SM each time.

Reply:
Ah, I didn't realize it would be thrown away.  Definitely need to lift it out then.